### ISSUE TITLE
Hierarchical resources and services

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/ballerina-spec#614]
Resources and services support hierarchy as follows. A service object has a collection of named resources. When the type of a resource is a service object, then this is semantically a subcollection. Listeners for protocols that have a concept of hierarchy (notably HTTP and GraphQL) can interpret this in the way that makes sense for that protocol.

This works similarly to filesystems. You can think of a service object as a directory and of a resource as an entry in the directory. A file hierarchy is represented by having directories with entries that are themselves directories. Filesystems typically provide a path syntax to make it convenient to work with the hierarchy.

We can provide some syntactic sugar to provide a similar convenience in Ballerina. We can allow

```
service object {
   resource function get foo/bar() returns int { ... }
}
```
as short for:

```
service object {
   final resource InferredType foo = service object {
      resource function get bar() returns int { ... }
   }
}
```

Objects get merged in the obvious way:

```
service object {
   resource function get foo/bar() returns int { ... }
   resource function get foo/baz() returns int { ... }
}
```

turns into

```
service object {
   final resource InferredType foo = service object {
      resource function get bar() returns int { ... }
      resource function get baz() returns int { ... }
   }
}
```

We also allow `.` to represent an empty path in the name of a resource. This represents the current service object as a resource. In the case of HTTP, each service object has a base path ending with a slash, and the HTTP Listener would map that base path onto the resource named `.`. For example, if a service object has a base path `/foo/`, then the HTTP Listener would map a path of `/foo/bar` onto the resource named bar and map a path of `/foo/` onto the resource named `.`.

The grammar is:

```
relative-resource-path := "." | (identifier ("/" identifier)*)
```

where both stored and computed resources use a relative-resource-path rather than an identifier as the name.

**Hierarchical service name**

We change the name in the service declaration to also be hierarchical. However, in this case, it is representing an absolute path, so the syntax is slightly different.

```
service-declaration := "service" [type-descriptor] [absolute-resource-path] "on" expression-list etc
absolute-resource-path := "/" | ("/" identifier)+
```

This means that there is no longer ambiguity between omitting the type descriptor and omitting the resource path.

The resource path in the service declaration is used only to provide the argument to attach; we eliminate the previous semantic of also defining a final variable.

The possible types of the argument to attach change accordingly:

*   if the absolute-resource-path is omitted, then the argument is nil;
*   if the absolute-resource-path is `/`, then the argument is an empty list;
*   otherwise, the absolute-resource-path is a list of one or more strings, one for each identifier in the absolute-resource-path

The Listener can control the allowed forms of the absolute-resource-path by declaring the appropriate type for the attach argument. For example, for Listeners that support only remote methods, typically the resource path will not make sense; in this case, the Listener can declare the type of the argument as ().

**Comment by jclark** [comment]
Note that since  we are planning for the HTTP listener to use remote methods to handle POST, names of remote methods on service objects should be allowed to be hierarchical, in addition to names of resources. But what does this mean for client? Maybe leave this for later.

**Comment by shafreenAnfar** [comment]
Historically there has been the concept of default resources. This concept was there in WSO2 ESB as well [1]. One common use-case of the default resource is to let user send custom 404/401 error messages when there is no resource found. User can define their own headers and payload content in the default resource and send it back. 

Though I haven't used it, the other possible use-case would be to capture these request and store it in some other location for auditing and debugging. 

This is supported in the current implementation as follows.

```ballerina
import ballerina/http;

service foo on new http:Listener(9090) {

    resource function baz(http:Caller caller, http:Request req) returns error? {
        // Some logic
    }

    @http:ResourceConfig {
        path: "/*"
    }
    resource function bar(http:Caller caller, http:Request req) returns error? {
        // Some logic
    }
}
```

In the above case, basically all the inbound requests except for localhost:9090/foo/baz resolves to `bar` resource.

[1] https://docs.wso2.com/display/ESB500/Working+with+APIs

**Comment by jclark** [comment]
We need to handle `self`  and service objects with state properly. Consider:

```
service object {
   int data = 3;
   resource function get a() returns int => self.data + 1;
   resource function get b/c() returns int => self.data + 2;
   resource function get b/d() returns int => self.data + 3;
}
```

The desugaring would need to be more like this:

```
service object {
   int data  = 3;
   resource function get a() returns int => self.data + 1;
   function __b_c() returns int => self.data + 2;
   function __b_d() returns int => self.data + 3;
   final resource InferredType b = let var oself = self in service object {
      final InferredType2 outer = oself;
      resource function get c() returns int  => self.outer.__b_c();
      resource function get d() returns int => self.outer.__b_d();
   }
}
```

But I suspect even this will interact purely with isolated.

This is so complicated, it may well be better not to explain this in terms of desugaring.

**Comment by sanjiva** [comment]
Couldn't agree more ;-).

Desugaring model can be within GraphQL services if needed. The HTTP case is much easier dealt with without that. I suggest that at the language level we just say the names can be hierarchical and indicate the syntax and leave it at that.

**Comment by jclark** [comment]
I have revised https://github.com/ballerina-platform/ballerina-spec/issues/607#issuecomment-721459883 to incorporate hierarchical resources in a first-class way, without desugaring. Support for hierarchical resources is now included in #607.
